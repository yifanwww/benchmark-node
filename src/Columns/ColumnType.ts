export enum ColumnType {
    FN,
    FN_ARGUMENT,
    PARAMETER,

    ERROR,
    ITERATIONS,
    MAX,
    MEAN,
    MEDIAN,
    MIN,
    OPS,
    Q1,
    Q3,
    STD_DEV,
    STD_ERR,

    CI_ERROR_L50,
    CI_ERROR_L70,
    CI_ERROR_L75,
    CI_ERROR_L80,
    CI_ERROR_L85,
    CI_ERROR_L90,
    CI_ERROR_L92,
    CI_ERROR_L95,
    CI_ERROR_L96,
    CI_ERROR_L97,
    CI_ERROR_L98,
    CI_ERROR_L99,
    CI_ERROR_L999,
    CI_ERROR_L9999,
}
