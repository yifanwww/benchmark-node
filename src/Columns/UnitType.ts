export enum UnitType {
    /** Dimensionless number. */
    Dimensionless,
    /** Dimensionless integer. */
    DimensionlessInteger,
    Time,
}
