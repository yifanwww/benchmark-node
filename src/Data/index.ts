export * from './ArgumentStore/ArgumentStore';
export * from './ArgumentStore/ArgumentStoreView';
export * from './Settings';
export * from './Statistics';
