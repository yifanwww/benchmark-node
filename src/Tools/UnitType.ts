export enum UnitType {
    /** Dimensionless number. */
    Dimensionless,
    /** Dimensionless integer. */
    DimensionlessInteger,
    Origin,
    Time,
}
