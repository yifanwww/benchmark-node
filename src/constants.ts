export const ANONYMOUS_FN_NAME = 'anonymous';
