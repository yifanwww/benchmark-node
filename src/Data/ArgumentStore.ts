import { range } from '../Tools/range';
import { LooseReadonlyArray } from '../types';
import { Arguments } from '../Parameterization/Arguments';
import { ArgumentStoreView } from './ArgumentStoreView';

export class ArgumentStore {
    private declare _argsList: readonly Arguments[];
    private declare _jitArgsList: readonly Arguments[];

    private declare _maxArgsLength: number;

    public get argsList() {
        return this._argsList;
    }

    public get jitArgsList() {
        return this._jitArgsList;
    }

    public get maxArgsLength() {
        return this._maxArgsLength;
    }

    public constructor(argsList: LooseReadonlyArray<Arguments> = [], jitArgsList: LooseReadonlyArray<Arguments> = []) {
        this._argsList = Array.isArray(argsList) ? argsList : [argsList];
        this._jitArgsList = [...(Array.isArray(jitArgsList) ? jitArgsList : [jitArgsList]), ...this._argsList];

        this._maxArgsLength = Math.max(this.getArgsLength(this._argsList), this.getArgsLength(this._jitArgsList));
    }

    private getArgsLength(argsArr: ReadonlyArray<Arguments>): number {
        let max = 0;
        for (const args of argsArr) {
            max = Math.max(max, args.args.length);
        }
        return max;
    }

    public *getViewEnumerator(): Generator<ArgumentStoreView, void> {
        if (this._argsList.length === 0) {
            yield new ArgumentStoreView(this, null);
        } else {
            for (const index of range(0, this._argsList.length)) {
                yield new ArgumentStoreView(this, index);
            }
        }
    }
}
