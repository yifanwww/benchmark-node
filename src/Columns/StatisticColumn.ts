import { Statistics } from '../Data';
import { UnitType } from '../Tools/UnitType';
import { BaseColumn } from './BaseColumn';
import { ColumnType } from './ColumnType';

export type Calc = (stats: Statistics) => number;

export class StatisticColumn extends BaseColumn<number> {
    private declare readonly _desc: string;
    private declare readonly _unitType: UnitType;

    get desc() {
        return this._desc;
    }

    get unitType() {
        return this._unitType;
    }

    constructor(type: ColumnType, name: string, desc: string, calc: Calc, unit: UnitType) {
        super(type, name, calc);

        this._desc = desc;
        this._unitType = unit;
    }
}
