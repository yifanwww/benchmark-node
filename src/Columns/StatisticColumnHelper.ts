import { Statistics } from '../Data';
import { Formatter } from '../Tools/Formatter';
import { TimeTool } from '../Tools/TimeTool';
import { TimeUnit, TimeUnitHelper } from '../Tools/TimeUnit';

import { StatisticColumn } from './StatisticColumn';
import { TableColumnHelper } from './TableColumnHelper';
import { UnitType } from './UnitType';

export class StatisticColumnHelper extends TableColumnHelper<number> {
    protected declare _column: StatisticColumn;

    private _timeUnit: TimeUnit = TimeUnit.NS;
    private _fractionDigit: number = 4;

    public override get column() {
        return this._column;
    }

    public get timeUnit() {
        return this._timeUnit;
    }

    public set timeUnit(value: TimeUnit) {
        this._timeUnit = value;
    }

    public setFractionDigit(fractionDigit: number): void {
        this._fractionDigit = fractionDigit;
    }

    public constructor(column: StatisticColumn) {
        super(column);
    }

    public findMinNumber(statsArr: Statistics[]): number {
        let min = Number.MAX_SAFE_INTEGER;

        for (const stats of statsArr) {
            const data = this._column.getData(stats);
            min = Math.min(min, data);
        }

        return min;
    }

    public findFractionDigit(statsArr: Statistics[]): void {
        let min = Number.MAX_SAFE_INTEGER;

        for (const _stats of statsArr) {
            const data = TimeTool.convert(this._column.getData(_stats), TimeUnit.NS, this._timeUnit);
            min = Math.min(min, data);
        }

        if (min <= 1) {
            this._fractionDigit = 4;
        } else if (min <= 10) {
            this._fractionDigit = 3;
        } else if (min <= 100) {
            this._fractionDigit = 2;
        } else if (min <= 1000) {
            this._fractionDigit = 1;
        } else {
            this._fractionDigit = 0;
        }
    }

    public override format(stats: Statistics): string {
        let never: never;
        switch (this._column.unit) {
            case UnitType.Dimensionless: {
                const data = this._column.getData(stats);
                return Formatter.beautifyNumber(data.toFixed(this._fractionDigit));
            }

            case UnitType.Time: {
                const data = TimeTool.convert(this._column.getData(stats), TimeUnit.NS, this._timeUnit);
                const num = Formatter.beautifyNumber(data.toFixed(this._fractionDigit));
                return `${num} ${TimeUnitHelper.getUnitStr(this._timeUnit)}`;
            }

            default:
                never = this._column.unit;
                return never;
        }
    }
}
