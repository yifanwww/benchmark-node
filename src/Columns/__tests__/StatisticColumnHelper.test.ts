import { Statistics } from '../../Data';
import { testConstants } from '../../TestTools/constants.test';
import { ColumnType } from '../ColumnType';
import { StatisticColumn } from '../StatisticColumn';
import { StatisticColumnHelper } from '../StatisticColumnHelper';
import { UnitType } from '../UnitType';

describe(`Test class \`${StatisticColumnHelper.name}\``, () => {
    const formatFnName = StatisticColumnHelper.prototype.format.name;

    it(`calls method ${formatFnName} with dimensionless statistic column`, () => {
        const stats = [new Statistics('name', testConstants.measurements.slice(), testConstants.ops)];

        const helper = new StatisticColumnHelper(
            new StatisticColumn(ColumnType.MEAN, 'name', 'desc', (_stats) => _stats.mean, UnitType.Dimensionless),
        );
        helper.findFractionDigit(stats);

        const output = helper.format(stats[0]);

        expect(output).toMatchSnapshot();
    });

    it(`calls method ${formatFnName} with dimensionless integer statistic column`, () => {
        const stats = [new Statistics('name', testConstants.measurements.slice(), testConstants.ops)];

        const helper = new StatisticColumnHelper(
            new StatisticColumn(ColumnType.MEAN, 'name', 'desc', (_stats) => _stats.n, UnitType.DimensionlessInteger),
        );
        helper.findFractionDigit(stats);

        const output = helper.format(stats[0]);

        expect(output).toMatchSnapshot();
    });

    it(`calls method ${formatFnName} with time statistic column`, () => {
        const stats = [new Statistics('name', testConstants.measurements.slice(), testConstants.ops)];

        const helper = new StatisticColumnHelper(
            new StatisticColumn(ColumnType.MEAN, 'name', 'desc', (_stats) => _stats.mean, UnitType.Time),
        );
        helper.findFractionDigit(stats);

        const output = helper.format(stats[0]);

        expect(output).toMatchSnapshot();
    });
});
