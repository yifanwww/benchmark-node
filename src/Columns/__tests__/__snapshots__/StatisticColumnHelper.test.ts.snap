// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test class \`StatisticColumnHelper\` calls method format with dimensionless integer statistic column 1`] = `"15"`;

exports[`Test class \`StatisticColumnHelper\` calls method format with dimensionless statistic column 1`] = `"0.2401"`;

exports[`Test class \`StatisticColumnHelper\` calls method format with time statistic column 1`] = `"0.2401 ns"`;
